### ISSUE TITLE
Cache backup - Return expired cache when cache expires 

### ISSUE THREAD

**Issue by bneiluj** [mpneuried/nodecache#62]
It would be great to have the possibility to define a backup for an expired cache. 

Basically, instead of listening on `cacheRegion.on( "expired", function(key, value )` and then update the cache when it expires. We could have an option defined like that: 

```
myCache.set( key, val, [ ttl, cacheBackup ], [callback] ) 
```

where `cacheBackup` is `true` if we want to have a backup and `false` otherwise. 

Workflow: 
- When the cache `expires` 
- And a `Retrieve a key (GET)` request is made
- It then return an `object` with the `key value pair` and a boolean `CacheBackup` equal to `true` = let the user know that it was an expired cache. 

It very useful when the http requests fail and the cache has expired.

**Comment by bneiluj** [comment]
This is my own workaround for now as I don't really want to hack into the coffee script code... :) 

```
'use strict';

var NodeCache = require('node-cache');

/**
 * {Object} params : params for node cache lib
 * Integer backupStdTTL : (default: 0) the standard ttl as number in seconds for every generated backup cache element.
 * String backupKeyExt : The string to search for the backup key. ie: "5151511-backup" where
 * 5151511 is the initial key and -backup is to tell it's a backup key
 */
class NCE extends NodeCache {
  constructor(
      params,
      backupStdTTL,
      backupKeyExt
  ) {
    super(params);
    this.backupStdTTL = backupStdTTL;
    this.backupKeyExt = backupKeyExt;
  }
  /**
   * Make sure a backup of the cache exist
   * when http fails and cache has expired
   */
  listenOnCache() {
    super.on( "expired", (key, value ) => {
      console.log("Cache expired, key: ", key);
      // If backup cache exist
      if (key.indexOf(this.backupKeyExt) != -1) {
        super.set(key, value, this.backupStdTTL);
      } else {
        // If no backup cache then create one
        super.set(key + this.backupKeyExt, value, this.backupStdTTL);
      }
    });
  }
}

// NCE = nodeCacheEtended
module.exports = NCE;

```

Then I call: 

`var NCE = require('./cache-backup-service.js');`
```
// Define node cache and nodeCacheExtended
var nodeCacheExtended = new NCE({stdTTL: 60,checkperiod: 30},60,"-backup");

// Listen on expired cache and create backup
nodeCacheExtended.listenOnCache();
```

**Comment by mpneuried** [comment]
Hi,
first, sorry for the late answer.

I looked at your request and your example code.

But ... i don't understand your case.
Why do you set the ttl of a key? The example will save the same data to the cache again after it was expired. So what's the advantage of this backup solution compared to `ttl=0` in fist place?

If you really need the behavior, i recommend to create your own npm module like `node-cache-backup` with the code above.
Please let me know, so i can add a link to your submodule to m the readme.

**Comment by bneiluj** [comment]
Hi, 

Most of the time when cache expires you want to retrieve a new version of the data and then update the cache, right ? But what happens if the data you are trying to fetch fails for any reasons ? Maybe the data server is down... 
In that case, the user tries to refresh the page but no data is found ! My solution is to have a backup of the case set up to unlimited time. So when the cache expires, it then tries to fetch data and if it fails, THEN it can use the backup cache. 
Basically, the problem is that when the cache expires you have no data to rely on... it's not good for reliable off internet applications and etc... 
You are right I can actually right a npm module for that but do you think my problem is actually a good user case ?

**Comment by mpneuried** [comment]
Ok, now i understand your scenario and in this case your solution is suitable.
But it's too specialized to implement it as a core feature of `node-cache`.

So, if you wrap your code into a npm module i will refer to it.
If you are not familiar with publishing modules, let me know.
I can also do this ;-)

**Comment by bneiluj** [comment]
Just saw your reply ! 
I'll create the npm module. I've been using this workaround for the past year on a high traffic applications actually :)
Keep in touch ! 👍
